### PR TITLE
navigateRegions: Convert to TypeScript

### PR DESCRIPTION
**Pull request by mirka**
Part of #35744

## What?

Converts the `navigateRegions` HOC to TypeScript.

## Testing Instructions

✅ Static checks pass.

### CODE REVIEW

File: packages/components/src/higher-order/navigate-regions/index.tsx [hunk 1]
```diff
@@ -33,27 +34,36 @@ const defaultShortcuts = {
...
+	const ref = useRef< HTMLDivElement >( null );
 	const [ isFocusingRegions, setIsFocusingRegions ] = useState( false );
 
-	function focusRegion( offset ) {
+	function focusRegion( offset: number ) {
 		const regions = Array.from(
-			ref.current.querySelectorAll( '[role="region"][tabindex="-1"]' )
+			ref.current?.querySelectorAll< HTMLElement >(
+				'[role="region"][tabindex="-1"]'
+			) ?? []
```
Comment (mirka) [inline packages/components/src/higher-order/navigate-regions/index.tsx:53]: Just making the fallback explicit here.

File: packages/components/src/higher-order/navigate-regions/index.tsx [hunk 2]
```diff
@@ -33,27 +34,36 @@ const defaultShortcuts = {
...
-			ref.current.ownerDocument.activeElement.closest(
+		const wrappingRegion =
+			ref.current?.ownerDocument?.activeElement?.closest< HTMLElement >(
 				'[role="region"][tabindex="-1"]'
-			)
-		);
+			);
+		const selectedIndex = wrappingRegion
+			? regions.indexOf( wrappingRegion )
+			: -1;
```
Comment (mirka) [inline packages/components/src/higher-order/navigate-regions/index.tsx:66]: Just making the fallback explicit, as `.indexOf()` will return -1 when `wrappingRegion` is undefined.

File: packages/keycodes/src/index.js [hunk 3]
```diff
@@ -42,7 +42,7 @@ import { isAppleOS } from './platform';
  *
  * @typedef {(character: string, isApple?: () => boolean) => T} WPKeyHandler
  */
-/** @typedef {(event: KeyboardEvent, character: string, isApple?: () => boolean) => boolean} WPEventKeyHandler */
+/** @typedef {(event: import('react').KeyboardEvent<HTMLElement>, character: string, isApple?: () => boolean) => boolean} WPEventKeyHandler */
```
Comment (mirka) [inline packages/keycodes/src/index.js]: I _think_ this is what we want, given that it's meant to handle React synthetic events? Let me know if you disagree.

File: packages/components/CHANGELOG.md [hunk 4]
```diff
@@ -9,6 +9,7 @@
 ### Internal
 
 -   `CircularOptionPicker`: Convert to TypeScript ([#47937](https://github.com/WordPress/gutenberg/pull/47937)).
+-   `navigateRegions` HOC: Convert to TypeScript ([#48632](https://github.com/WordPress/gutenberg/pull/48632)).
```
Comment (ciampo) [inline packages/components/CHANGELOG.md]: We'll need a rebase to move this entry to the new Unreleased section

File: packages/components/src/higher-order/navigate-regions/index.tsx [hunk 5]
```diff
@@ -33,27 +34,36 @@ const defaultShortcuts = {
...
 		);
 		if ( ! regions.length ) {
 			return;
 		}
 		let nextRegion = regions[ 0 ];
 		// Based off the current element, use closest to determine the wrapping region since this operates up the DOM. Also, match tabindex to avoid edge cases with regions we do not want.
-		const selectedIndex = regions.indexOf(
-			ref.current.ownerDocument.activeElement.closest(
+		const wrappingRegion =
+			ref.current?.ownerDocument?.activeElement?.closest< HTMLElement >(
```
Comment (ciampo) [inline packages/components/src/higher-order/navigate-regions/index.tsx:61]: Gotta love optional chaining ❓ ⛓️

File: packages/keycodes/src/index.js (hunk 3 above)
Comment (ciampo) [reply packages/keycodes/src/index.js]: Mhh, this is a tricky one.

For sure, the `onKeyDown` property returned by `useNavigateRegions` needs to take a `React.KeyboardEvent< HTMLDivElement >` as its argument.

Personally, I'm almost tempted to introduce a runtime change to the `navigateRegions` component and avoid making changes to the `@wordpress/keycodes` package, since from a quick look it seems that the package is not assuming react as a web dependency (only `react-native` is mentioned).

<details>
<summary>With this in mind, as an alternative approach we could use the `nativeEvent` property in `useNavigateRegions`</summary>

```diff
diff --git a/packages/components/src/higher-order/navigate-regions/index.tsx b/packages/components/src/higher-order/navigate-regions/index.tsx
index f2b9c31bfb..44afd39f4c 100644
--- a/packages/components/src/higher-order/navigate-regions/index.tsx
+++ b/packages/components/src/higher-order/navigate-regions/index.tsx
@@ -96,13 +96,19 @@ export function useNavigateRegions( shortcuts: Shortcuts = defaultShortcuts ) {
 		onKeyDown( event: React.KeyboardEvent< HTMLDivElement > ) {
 			if (
 				shortcuts.previous.some( ( { modifier, character } ) => {
-					return isKeyboardEvent[ modifier ]( event, character );
+					return isKeyboardEvent[ modifier ](
+						event.nativeEvent,
+						character
+					);
 				} )
 			) {
 				focusRegion( -1 );
 			} else if (
 				shortcuts.next.some( ( { modifier, character } ) => {
-					return isKeyboardEvent[ modifier ]( event, character );
+					return isKeyboardEvent[ modifier ](
+						event.nativeEvent,
+						character
+					);
 				} )
 			) {
 				focusRegion( 1 );
diff --git a/packages/keycodes/src/index.js b/packages/keycodes/src/index.js
index 15e96dca77..cd31fef5fa 100644
--- a/packages/keycodes/src/index.js
+++ b/packages/keycodes/src/index.js
@@ -42,7 +42,7 @@ import { isAppleOS } from './platform';
  *
  * @typedef {(character: string, isApple?: () => boolean) => T} WPKeyHandler
  */
-/** @typedef {(event: import('react').KeyboardEvent<HTMLElement>, character: string, isApple?: () => boolean) => boolean} WPEventKeyHandler */
+/** @typedef {(event: KeyboardEvent, character: string, isApple?: () => boolean) => boolean} WPEventKeyHandler */
 
 /** @typedef {( isApple: () => boolean ) => WPModifierPart[]} WPModifier */
 
@@ -346,7 +346,7 @@ export const shortcutAriaLabel = mapValues(
  * From a given KeyboardEvent, returns an array of active modifier constants for
  * the event.
  *
- * @param {import('react').KeyboardEvent<HTMLElement>} event Keyboard event.
+ * @param {KeyboardEvent} event Keyboard event.
  *
  * @return {Array<WPModifierPart>} Active modifier constants.
  */

```

</details>

Basically, trading off a small runtime change in exchange for not changing another package's APIs. What do you think?

Comment (mirka) [reply packages/keycodes/src/index.js]: Thanks for digging deeper here! I looked into the codebase to see how `isKeyboardEvent()` is being used, and it is in fact handling both [native](https://github.com/WordPress/gutenberg/blob/6be1e861fbf7049935a3733a1d2061d360c079f9/packages/block-editor/src/components/rich-text/use-remove-browser-shortcuts.js#L24) and [synthetic](https://github.com/WordPress/gutenberg/blob/6be1e861fbf7049935a3733a1d2061d360c079f9/packages/block-library/src/navigation-submenu/edit.js#L293) events. The synthetic usages would all be errors if they were typechecked. So I think it would be best to correct the types so it matches the actual usage 👉 b94985a6

How does that sound?

File: packages/components/CHANGELOG.md (hunk 4 above)
Comment (mirka) [reply packages/components/CHANGELOG.md]: ✅ 3e8e7f16

File: packages/keycodes/src/index.js (hunk 3 above)
Comment (ciampo) [reply packages/keycodes/src/index.js]: Sounds good to me!

**Review by ciampo — APPROVED** [review]
🚀
